File: src/main/scala/com/nthportal/shell/async/InputAction.scala
package com.nthportal.shell
package async

import scala.concurrent.{Future, Promise}
import scala.util.Try

/**
  * An action requested to be executed by a [[Shell]] by the
  * [[InputProvider input source]] of an [[AsyncShell]].
  *
  * @tparam T the type of the result of the action
  */
trait InputAction[T] {
  private val promise = Promise[T]

  /**
    * Returns a [[Future]] which will contain the result of this action.
    *
    * @return a Future which will contain the result of this action
    */
  def future: Future[T] = promise.future

  /**
    * Performs this action using a shell, and makes the result
    * available in [[future]]. Performs the action defined as [[action]].
    *
    * @param shell the shell with which to perform the action
    */
  private[async] def doAction(implicit shell: Shell): Unit = promise.complete(Try(action(shell)))

  /**
    * Returns the result of an action performed with a [[Shell]].
    *
    * @param shell the shell with which to perform the action
    * @return the result of the action
    */
  private[InputAction] def action(shell: Shell): T
}

object InputAction {
  /**
    * Creates an input action for tab-completing a line.
    *
    * @param line the line to be tab-completed
    * @return an input action for tab-completing the given line
    * @see [[Shell.tabComplete]]
    */
  def tabCompletion(line: String): InputAction[ImmutableSeq[String]] = _.tabComplete(line)

  /**
    * Creates an input action for executing a line.
    *
    * @param line the line to be executed
    * @return an input action for executing the given line
    * @see [[Shell.executeLine]]
    */
  def execution(line: String): InputAction[Unit] = _.executeLine(line)
}

File: src/main/java/com/nthportal/shell/compat/async/InputAction.java
package com.nthportal.shell.compat.async;

import com.nthportal.shell.compat.Shell;
import scala.compat.java8.FutureConverters;
import scala.concurrent.Future;

import java.util.List;
import java.util.concurrent.CompletionStage;

/**
 * An action requested to be executed by a {@link Shell} by the
 * {@link InputProvider input source} of an {@link AsyncShell}.
 *
 * @param <T> the type of the result of the action
 */
public interface InputAction<T> {
    /**
     * Returns a {@link CompletionStage} which will contain the result of this action.
     *
     * @return a CompletionStage which will contain the result of this action
     */
    default CompletionStage<T> completionStage() {
        return FutureConverters.toJava(future());
    }

    /**
     * Returns a {@link Future} which will contain the result of this action.
     *
     * @return a Future which will contain the result of this action
     */
    Future<T> future();

    /**
     * Returns the result of an action performed with a {@link Shell}.
     *
     * @param shell the shell with which to perform the action
     * @return the result of the action
     */
    T action(Shell shell);

    /**
     * Creates an input action for tab-completing a line.
     *
     * @param line the line to be tab-completed
     * @return an input action for tab-completing the given line
     * @see Shell#tabComplete(String)
     */
    static InputAction<List<String>> tabCompletion(String line) {
        return JCompatInputAction.tabCompletion(line);
    }

    /**
     * Creates an input action for executing a line.
     *
     * @param line the line to be executed
     * @return an input action for executing the given line
     * @see Shell#executeLine(String)
     */
    static InputAction<Void> execution(String line) {
        return JCompatInputAction.execution(line);
    }
}
